### PR TITLE
runtime-rs: agent implements:

### PR DESCRIPTION
**Pull request by quanweiZhou**
Responsible for communicating with the agent, such as kata-agent in the VM

Fixes: #3785
Signed-off-by: Quanwei Zhou <quanweiZhou@linux.alibaba.com>

### CODE REVIEW

File: src/runtime-rs/crates/agent/src/vsock.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                tokio::time::sleep(std::time::Duration::from_millis(config.dial_timeout_ms)).await;
+                continue;
+            }
+        }
+    }
+    Err(anyhow!("cannot connect to agent ttrpc server"))
+}
+
+async fn connect_helper(config: &VsockConnectConfig) -> Result<UnixStream> {
+    info!(sl!(), "connect vsock {:?}", &config);
```
Comment (bergwolf) [inline src/runtime-rs/crates/agent/src/vsock.rs:63]: connect config, or config.uds?

File: src/runtime-rs/crates/agent/src/vsock.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    let mut stream = UnixStream::connect(&config.uds).await.context("connect")?;
```
Comment (bergwolf) [inline src/runtime-rs/crates/agent/src/vsock.rs:64]: So it only supports hybrid vsock that clh/firecracker has? For qemu, we still need to connect to vhost vsock that is of format <cid:port>.

File: src/libs/kata-types/src/config/agent.rs [hunk 3]
```diff
@@ -27,15 +27,29 @@ pub struct Agent {
     #[serde(default)]
     pub enable_tracing: bool,
 
-    /// Enable debug console.
-    /// If enabled, user can connect guest OS running inside hypervisor through
-    /// "kata-runtime exec <sandbox-id>" command
+    /// Agent server port
     #[serde(default)]
-    pub debug_console_enabled: bool,
```
Comment (liubin) [inline src/libs/kata-types/src/config/agent.rs:34]: debug console is intentionally deleted?

File: src/runtime-rs/crates/agent/src/kata/mod.rs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    pub port: u32,
+}
+
+pub(crate) struct KataAgentInner {
+    pub client: Option<Client>,
+    pub client_fd: RawFd,
+    pub uds: String,
+
+    /// Agent server port
+    pub server_port: u32,
```
Comment (liubin) [inline src/runtime-rs/crates/agent/src/kata/mod.rs:36]: These fields are the same as `Agent` struct in `src/libs/kata-types/src/config/agent.rs`, so can save `Agent` as a field in `KataAgentInner`?

File: src/runtime-rs/crates/agent/src/log_forwarder.rs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+    pub(crate) fn stop(&mut self) {
+        let task_handler = self.task_handler.take();
+        if let Some(handler) = task_handler {
+            handler.abort();
+            info!(sl!(), "abort log forwarder thread");
+        }
+    }
+
+    // start_agent_log_forwarder connect kata-agent log vsock and copy data to hypervisor's log stream
```
Comment (liubin) [inline src/runtime-rs/crates/agent/src/log_forwarder.rs:37]: Nit: `start_agent_log_forwarder` -> `start` ?

File: src/runtime-rs/crates/agent/src/vsock.rs (hunk 1 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/agent/src/vsock.rs:63]: Thanks for your review, change to `ConnectConfig`.

File: src/runtime-rs/crates/agent/src/vsock.rs (hunk 2 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/agent/src/vsock.rs:64]: Thanks for your review,  update to support both `vsock` and `hvsock`.

File: src/libs/kata-types/src/config/agent.rs (hunk 3 above)
Comment (quanweiZhou) [reply src/libs/kata-types/src/config/agent.rs:34]: Thanks for your review, sorry for accidentally deleting it has restored it.

File: src/runtime-rs/crates/agent/src/kata/mod.rs (hunk 4 above)
Comment (quanweiZhou) [reply src/runtime-rs/crates/agent/src/kata/mod.rs:36]: Thanks for your review, change to use AgentConfig as a field in KataAgentInner.

**Review by liubin — APPROVED** [review]

File: src/runtime-rs/crates/agent/src/sock/mod.rs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+#[async_trait]
+pub trait Sock: Send + Sync {
+    async fn connect(&self, config: &ConnectConfig) -> Result<Stream>;
+}
+
+// Supported sock address formats are:
+//   - vsock://<cid>:<port>
+//   - hvsock://<path>:<port>. Firecracker implements the virtio-vsock device
```
Comment (bergwolf) [inline src/runtime-rs/crates/agent/src/sock/mod.rs:109]: Nit: vsock is supported by both firecracker and cloud-hypervisor.

**Review by bergwolf — APPROVED** [review]
lgtm!

**Pull request merged**